[BUG] Indexer_personal_info terminates abnormally during Quorum failover

**Describe the bug**

- Indexer_personal_info.py terminates abnormally during Quorum failover.
- The following error occurs.

```
[2021-08-08 16:00:53 +0000] [INDEXER-Personal-Info] [37] [ERROR] HTTPConnectionPool(host='quorumazc.xxxxx.xxxxx.local', port=8545): Max retries exceeded with url: / (Caused by NewConnectionError('<urllib3.connection.HTTPConnection object at 0x7f11d324d5b0>: Failed to establish a new connection: [Errno 111] Connection refused'))
Traceback (most recent call last):
  File "/home/apl/.pyenv/versions/3.8.7/lib/python3.8/site-packages/urllib3/connection.py", line 169, in _new_conn
    conn = connection.create_connection(
  File "/home/apl/.pyenv/versions/3.8.7/lib/python3.8/site-packages/urllib3/util/connection.py", line 96, in create_connection
    raise err
  File "/home/apl/.pyenv/versions/3.8.7/lib/python3.8/site-packages/urllib3/util/connection.py", line 86, in create_connection
    sock.connect(sa)
ConnectionRefusedError: [Errno 111] Connection refused
During handling of the above exception, another exception occurred:
Traceback (most recent call last):
  File "/home/apl/.pyenv/versions/3.8.7/lib/python3.8/site-packages/urllib3/connectionpool.py", line 699, in urlopen
    httplib_response = self._make_request(
  File "/home/apl/.pyenv/versions/3.8.7/lib/python3.8/site-packages/urllib3/connectionpool.py", line 394, in _make_request
    conn.request(method, url, **httplib_request_kw)
  File "/home/apl/.pyenv/versions/3.8.7/lib/python3.8/site-packages/urllib3/connection.py", line 234, in request
    super(HTTPConnection, self).request(method, url, body=body, headers=headers)
  File "/home/apl/.pyenv/versions/3.8.7/lib/python3.8/http/client.py", line 1255, in request
    self._send_request(method, url, body, headers, encode_chunked)
  File "/home/apl/.pyenv/versions/3.8.7/lib/python3.8/http/client.py", line 1301, in _send_request
    self.endheaders(body, encode_chunked=encode_chunked)
  File "/home/apl/.pyenv/versions/3.8.7/lib/python3.8/http/client.py", line 1250, in endheaders
    self._send_output(message_body, encode_chunked=encode_chunked)
  File "/home/apl/.pyenv/versions/3.8.7/lib/python3.8/http/client.py", line 1010, in _send_output
    self.send(msg)
  File "/home/apl/.pyenv/versions/3.8.7/lib/python3.8/http/client.py", line 950, in send
    self.connect()
  File "/home/apl/.pyenv/versions/3.8.7/lib/python3.8/site-packages/urllib3/connection.py", line 200, in connect
    conn = self._new_conn()
  File "/home/apl/.pyenv/versions/3.8.7/lib/python3.8/site-packages/urllib3/connection.py", line 181, in _new_conn
    raise NewConnectionError(
urllib3.exceptions.NewConnectionError: <urllib3.connection.HTTPConnection object at 0x7f11d324d5b0>: Failed to establish a new connection: [Errno 111] Connection refused
During handling of the above exception, another exception occurred:
Traceback (most recent call last):
  File "/home/apl/.pyenv/versions/3.8.7/lib/python3.8/site-packages/requests/adapters.py", line 439, in send
    resp = conn.urlopen(
  File "/home/apl/.pyenv/versions/3.8.7/lib/python3.8/site-packages/urllib3/connectionpool.py", line 755, in urlopen
    retries = retries.increment(
  File "/home/apl/.pyenv/versions/3.8.7/lib/python3.8/site-packages/urllib3/util/retry.py", line 574, in increment
    raise MaxRetryError(_pool, url, error or ResponseError(cause))
urllib3.exceptions.MaxRetryError: HTTPConnectionPool(host='quorumazc.xxxxx.xxxxx.local', port=8545): Max retries exceeded with url: / (Caused by NewConnectionError('<urllib3.connection.HTTPConnection object at 0x7f11d324d5b0>: Failed to establish a new connection: [Errno 111] Connection refused'))
During handling of the above exception, another exception occurred:
Traceback (most recent call last):
  File "batch/indexer_personal_info.py", line 231, in <module>
    processor.process()
  File "batch/indexer_personal_info.py", line 110, in process
    self.__refresh_personal_info_list()
  File "batch/indexer_personal_info.py", line 132, in __refresh_personal_info_list
    personal_info_address = token_contract.functions.personalInfoAddress().call()
  File "/home/apl/.pyenv/versions/3.8.7/lib/python3.8/site-packages/web3/contract.py", line 954, in call
    return call_contract_function(
  File "/home/apl/.pyenv/versions/3.8.7/lib/python3.8/site-packages/web3/contract.py", line 1499, in call_contract_function
    return_data = web3.eth.call(call_transaction, block_identifier=block_id)
  File "/home/apl/.pyenv/versions/3.8.7/lib/python3.8/site-packages/web3/module.py", line 58, in caller
    result = w3.manager.request_blocking(method_str, params, error_formatters)
  File "/home/apl/.pyenv/versions/3.8.7/lib/python3.8/site-packages/web3/manager.py", line 154, in request_blocking
    response = self._make_request(method, params)
  File "/home/apl/.pyenv/versions/3.8.7/lib/python3.8/site-packages/web3/manager.py", line 133, in _make_request
    return request_func(method, params)
  File "cytoolz/functoolz.pyx", line 250, in cytoolz.functoolz.curry.__call__
  File "/home/apl/.pyenv/versions/3.8.7/lib/python3.8/site-packages/web3/middleware/formatting.py", line 76, in apply_formatters
    response = make_request(method, params)
  File "/home/apl/.pyenv/versions/3.8.7/lib/python3.8/site-packages/web3/middleware/gas_price_strategy.py", line 34, in middleware
    return make_request(method, params)
  File "cytoolz/functoolz.pyx", line 250, in cytoolz.functoolz.curry.__call__
  File "/home/apl/.pyenv/versions/3.8.7/lib/python3.8/site-packages/web3/middleware/formatting.py", line 74, in apply_formatters
    response = make_request(method, formatted_params)
  File "/home/apl/.pyenv/versions/3.8.7/lib/python3.8/site-packages/web3/middleware/attrdict.py", line 33, in middleware
    response = make_request(method, params)
  File "cytoolz/functoolz.pyx", line 250, in cytoolz.functoolz.curry.__call__
  File "/home/apl/.pyenv/versions/3.8.7/lib/python3.8/site-packages/web3/middleware/formatting.py", line 74, in apply_formatters
    response = make_request(method, formatted_params)
  File "/home/apl/.pyenv/versions/3.8.7/lib/python3.8/site-packages/web3/middleware/normalize_errors.py", line 25, in middleware
    result = make_request(method, params)
  File "cytoolz/functoolz.pyx", line 250, in cytoolz.functoolz.curry.__call__
  File "/home/apl/.pyenv/versions/3.8.7/lib/python3.8/site-packages/web3/middleware/formatting.py", line 74, in apply_formatters
    response = make_request(method, formatted_params)
  File "cytoolz/functoolz.pyx", line 250, in cytoolz.functoolz.curry.__call__
  File "/home/apl/.pyenv/versions/3.8.7/lib/python3.8/site-packages/web3/middleware/formatting.py", line 74, in apply_formatters
    response = make_request(method, formatted_params)
  File "cytoolz/functoolz.pyx", line 250, in cytoolz.functoolz.curry.__call__
  File "/home/apl/.pyenv/versions/3.8.7/lib/python3.8/site-packages/web3/middleware/formatting.py", line 76, in apply_formatters
    response = make_request(method, params)
  File "/home/apl/.pyenv/versions/3.8.7/lib/python3.8/site-packages/web3/middleware/exception_retry_request.py", line 104, in middleware
    return make_request(method, params)
  File "batch/../app/utils/web3_utils.py", line 112, in make_request
    return super().make_request(method, params)
  File "/home/apl/.pyenv/versions/3.8.7/lib/python3.8/site-packages/web3/providers/rpc.py", line 92, in make_request
    raw_response = make_post_request(
  File "/home/apl/.pyenv/versions/3.8.7/lib/python3.8/site-packages/web3/_utils/request.py", line 39, in make_post_request
    response = session.post(endpoint_uri, data=data, *args, **kwargs)  # type: ignore
  File "/home/apl/.pyenv/versions/3.8.7/lib/python3.8/site-packages/requests/sessions.py", line 590, in post
    return self.request('POST', url, data=data, json=json, **kwargs)
  File "/home/apl/.pyenv/versions/3.8.7/lib/python3.8/site-packages/requests/sessions.py", line 542, in request
    resp = self.send(prep, **send_kwargs)
  File "/home/apl/.pyenv/versions/3.8.7/lib/python3.8/site-packages/requests/sessions.py", line 655, in send
    r = adapter.send(request, **kwargs)
  File "/home/apl/.pyenv/versions/3.8.7/lib/python3.8/site-packages/requests/adapters.py", line 516, in send
    raise ConnectionError(e, request=request)
requests.exceptions.ConnectionError: HTTPConnectionPool(host='quorumazc.xxxxx.xxxxx.local', port=8545): Max retries exceeded with url: / (Caused by NewConnectionError('<urllib3.connection.HTTPConnection object at 0x7f11d324d5b0>: Failed to establish a new connection: [Errno 111] Connection refused'))
```

**Expected behavior**
- When a failover occurs, the Quorum to be connected is automatically switched and the process continues.

--- a/app/utils/web3_utils.py
+++ b/app/utils/web3_utils.py
@@ -19,12 +19,10 @@ SPDX-License-Identifier: Apache-2.0
 import sys
 import threading
 from typing import Any
+import time
 
 from sqlalchemy import create_engine
-from sqlalchemy.orm import (
-    sessionmaker,
-    scoped_session
-)
+from sqlalchemy.orm import Session
 from web3 import Web3
 from web3.middleware import geth_poa_middleware
 from web3.types import (
@@ -32,19 +30,19 @@ from web3.types import (
     RPCResponse
 )
 from eth_typing import URI
+from requests.exceptions import ConnectionError
 
 from config import (
     WEB3_HTTP_PROVIDER,
     DATABASE_URL,
-    DB_ECHO
+    DB_ECHO,
+    WEB3_REQUEST_RETRY_COUNT,
+    WEB3_REQUEST_WAIT_TIME
 )
 from app.model.db import Node
 from app.exceptions import ServiceUnavailableError
 
 engine = create_engine(DATABASE_URL, echo=DB_ECHO)
-db_session = scoped_session(sessionmaker())
-db_session.configure(bind=engine)
-
 thread_local = threading.local()
 
 
@@ -91,32 +89,49 @@ class FailOverHTTPProvider(Web3.HTTPProvider):
         self.endpoint_uri = None
 
     def make_request(self, method: RPCEndpoint, params: Any) -> RPCResponse:
+        db_session = Session(autocommit=False, autoflush=True, bind=engine)
+        try:
+            if FailOverHTTPProvider.is_default is not None and not FailOverHTTPProvider.is_default:
+                counter = 0
+                while counter <= WEB3_REQUEST_RETRY_COUNT:
+                    # Switch alive node
+                    _node = db_session.query(Node). \
+                        filter(Node.is_synced == True). \
+                        order_by(Node.priority). \
+                        order_by(Node.id). \
+                        first()
+                    if _node is None:
+                        counter += 1
+                        if counter <= WEB3_REQUEST_RETRY_COUNT:
+                            time.sleep(WEB3_REQUEST_WAIT_TIME)
+                            continue
+                        raise ServiceUnavailableError("block synchronization is down")
+                    self.endpoint_uri = URI(_node.endpoint_uri)
 
-        if FailOverHTTPProvider.is_default is not None and not FailOverHTTPProvider.is_default:
-            # Switch alive node
-            _node = db_session.query(Node). \
-                filter(Node.is_synced == True). \
-                order_by(Node.priority). \
-                order_by(Node.id). \
-                first()
-            db_session.rollback()
-            if _node is None:
-                raise ServiceUnavailableError("block synchronization is down")
-            self.endpoint_uri = URI(_node.endpoint_uri)
-        else:
-            # Use default(primary) node
-            self.endpoint_uri = URI(WEB3_HTTP_PROVIDER)
-            FailOverHTTPProvider.set_is_default()
+                    # Call RPC method
+                    try:
+                        return super().make_request(method, params)
+                    except ConnectionError:
+                        counter += 1
+                        if counter <= WEB3_REQUEST_RETRY_COUNT:
+                            time.sleep(WEB3_REQUEST_WAIT_TIME)
+                            continue
+                        raise ServiceUnavailableError("block synchronization is down")
+            else:
+                # Use default(primary) node
+                self.endpoint_uri = URI(WEB3_HTTP_PROVIDER)
+                FailOverHTTPProvider.set_is_default(db_session)
 
-        # Call RPC method
-        return super().make_request(method, params)
+                # Call RPC method
+                return super().make_request(method, params)
+        finally:
+            db_session.close()
 
     @staticmethod
-    def set_is_default():
+    def set_is_default(db_session: Session):
         node = db_session.query(Node). \
             filter(Node.endpoint_uri != None). \
             first()
-        db_session.rollback()
         if node is None:
             FailOverHTTPProvider.is_default = True
         else:
@@ -125,7 +140,11 @@ class FailOverHTTPProvider(Web3.HTTPProvider):
 
 # NOTE: Loaded module before DB table create when executed from Pytest.
 if "pytest" not in sys.modules:
-    # First loaded module
-    if FailOverHTTPProvider.is_default is None:
-        # If never running the block monitor processor, use default(primary) node.
-        FailOverHTTPProvider.set_is_default()
+    db_session = Session(autocommit=False, autoflush=True, bind=engine)
+    try:
+        # First loaded module
+        if FailOverHTTPProvider.is_default is None:
+            # If never running the block monitor processor, use default(primary) node.
+            FailOverHTTPProvider.set_is_default(db_session)
+    finally:
+        db_session.close()

--- a/config.py
+++ b/config.py
@@ -105,7 +105,7 @@ if "pytest" in sys.modules:  # for unit test
         if os.environ.get("E2EE_RSA_RESOURCE_MODE") else 0
     E2EE_RSA_RESOURCE = os.environ.get("E2EE_RSA_RESOURCE") or "tests/data/rsa_private.pem"
     E2EE_RSA_PASSPHRASE = os.environ.get("E2EE_RSA_PASSPHRASE") or "password"
-elif "alembic" in sys.modules or "manage.py" in sys.argv[0]:  # for migration
+elif "alembic" in sys.modules or "manage.py" in sys.argv[0] or APP_ENV == "local":  # for migration or local
     E2EE_RSA_RESOURCE_MODE = int(os.environ.get("E2EE_RSA_RESOURCE_MODE")) \
         if os.environ.get("E2EE_RSA_RESOURCE_MODE") else 0
     E2EE_RSA_RESOURCE = os.environ.get("E2EE_RSA_RESOURCE") or "tests/data/rsa_private.pem"
@@ -155,7 +155,13 @@ if APP_ENV == "local":
 else:
     BLOCK_GENERATION_SPEED_THRESHOLD = int(os.environ.get("BLOCK_GENERATION_SPEED_THRESHOLD")) \
         if os.environ.get("BLOCK_GENERATION_SPEED_THRESHOLD") else 20
-# auto transfer approval interval(second)
+WEB3_REQUEST_RETRY_COUNT = int(os.environ.get("WEB3_REQUEST_RETRY_COUNT")) if os.environ.get(
+    "WEB3_REQUEST_RETRY_COUNT") else 3
+WEB3_REQUEST_WAIT_TIME = int(os.environ.get("WEB3_REQUEST_WAIT_TIME")) \
+    if os.environ.get("WEB3_REQUEST_WAIT_TIME") else BLOCK_SYNC_STATUS_SLEEP_INTERVAL  # Same batch interval
+
+# Auto transfer approval
+# interval(second)
 AUTO_TRANSFER_APPROVAL_INTERVAL = int(os.environ.get("AUTO_TRANSFER_APPROVAL_INTERVAL")) \
     if os.environ.get("AUTO_TRANSFER_APPROVAL_INTERVAL") else 10
 

